Add priority for Highlight

jesxilin opened the pull request:
#### 3b510397ce89a479d6d644aa21183a8de3282ec2
<pre>
Add priority for Highlight
<a href="https://bugs.webkit.org/show_bug.cgi?id=257611">https://bugs.webkit.org/show_bug.cgi?id=257611</a>
rdar://110125853

Reviewed by Megan Gardner.

Included implementation for priority and insertion order of css highlights.
Fixed selection inconsistency to always have active selection painted the same, above all markedTexts.
Fixed painting foreground color text so non-clashing styles of lower priority are still painted.
Added implementation for repainting when priority changes.
Added implementation for painting same semi-transparent overlapping highlight to be painted as a union.
Added a check if users assigned text color to take text color of prioritized highlight.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority.html: Added.
* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::Highlight::setPriority):
* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::Highlight::setPriority): Deleted.
* Source/WebCore/Modules/highlight/HighlightRegister.cpp:
(WebCore::HighlightRegister::setFromMapLike):
(WebCore::HighlightRegister::clear):
(WebCore::HighlightRegister::remove):
* Source/WebCore/Modules/highlight/HighlightRegister.h:
(WebCore::HighlightRegister::highlightNames const):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::subdivide):
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/MarkedText.h:
(WebCore::MarkedText::MarkedText):
(WebCore::MarkedText::isHashTableDeletedValue const):
(WTF::add):
(WTF::HashTraits&lt;WebCore::MarkedText&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebCore::MarkedText&gt;::isDeletedValue):
(WTF::DefaultHash&lt;WebCore::MarkedText&gt;::hash):
(WTF::DefaultHash&lt;WebCore::MarkedText&gt;::equal):
(WebCore::MarkedText::operator== const): Deleted.
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
(WebCore::coalesceAdjacentWithSameRanges):
(WebCore::orderHighlights):
(WebCore::StyledMarkedText::subdivideAndResolve):
(WebCore::StyledMarkedText::coalesceAdjacentWithEqualDecorations):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForegroundAndDecorations):
* Source/WebCore/rendering/TextPaintStyle.h:

Canonical link: <a href="https://commits.webkit.org/265812@main">https://commits.webkit.org/265812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39f1fe938744be9320fdc26828e227d438bf90d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11876 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12449 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13521 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11895 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14464 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12056 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14172 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10053 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13943 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10786 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17917 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11234 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10941 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14110 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11351 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9385 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10665 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2925 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14809 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -103,5 +103,11 @@ void Highlight::repaint()
 
+void Highlight::setPriority(int priority)
+{
+    if (priority >= 0)
```
maybe this can be an `ASSERT` instead of / in addition to this check?

aprotyas replied:
Yeah, this shouldn't be silent I think

jesxilin replied:
would it look something like this `ASSERT(!priority);`?

rr-codes replied:
It would be ASSERT(priority >= 0). But you should probably address Aditya’s comment about if negative priorities are actually allowed.

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,42 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                    if (highlightStart < highlightEnd) {
+                        // if we can just append it to the end, do that instead
+                        if (markedTexts.size() == 0
+                            || markedTexts[markedTexts.size()-1].priority <= highlightRegister->map().get(highlightName)->priority()) {
```
I think you can do `markedTexts.isEmpty() || markedTexts.last().priority ...`

aprotyas replied:
nit

```diff
- markedTexts[markedTexts.size()-1].priority
+ markedTexts.last().priority
```

rr-codes commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,42 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                                    inserted = true;
+                                    break;
+                                }
+                                if (markedTexts[index].priority == highlightRegister->map().get(highlightName)->priority()) {
```
Could probably save `highlightRegister->map().get(highlightName)->priority)(` into a variable and just re-use that variable instead of repeating this call chain

rr-codes commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,17 +68,33 @@ struct MarkedText {
     AtomString highlightName { };
-
+    int priority { 0 };
+    bool m_isHashTableDeletedValue { false };
```
this should probably be private

jesxilin replied:
Did you mean priority should be private or the bool? Or both?

rr-codes replied:
just `m_isHashTableDeletedValue`, since clients of this struct really shouldn't know or care about it

rr-codes commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -88,4 +104,26 @@ struct MarkedText {
+    }
+
+    static constexpr bool safeToCompareToEmptyOrDeleted = true;
+    // static constexpr bool hasHashInValue = true; // This is not correct, but for debugging of RobinHoodHashSet.
```
should remove this comment

whsieh commented on Source/WebCore/Modules/highlight/HighlightRegister.h:
```diff
@@ -55,10 +55,12 @@ class HighlightRegister : public RefCounted<HighlightRegister> {
 private:
     HighlightRegister() = default;
     HashMap<AtomString, Ref<Highlight>> m_map;
+    ListHashSet<AtomString> m_set; // holds highlight names in order
```
Nit - perhaps, this could be named something like `m_highlightNamesInOrder` or just `m_orderedHighlightNames`? (And then we could remove the comment).

We could also just use a `Vector` here, since we already have `m_map` for fast `contains()` checks.

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,22 +143,52 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        // Sort so highlights follow correct priority/insertion order
+        if (!markedTexts[0].highlightName.isNull()) {
+            // Maintain copy of old unordered markedTexts
+            Vector<MarkedText> dummy;
```
should probably have a more descriptive name

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,22 +143,52 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+            Vector<MarkedText> dummy;
+            dummy = markedTexts;
+            markedTexts.clear();
+            for (auto& hiliteName : markedTextsNames) {
```
`hiliteName` -> `highlightName`

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,22 +143,52 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
     // Compute frontmost overlapping styled marked texts.
-    frontmostMarkedTexts.reserveInitialCapacity(markedTexts.size());
-    frontmostMarkedTexts.uncheckedAppend(resolveStyleForMarkedText(markedTexts[0], baseStyle, renderer, lineStyle, paintInfo));
+    frontmostMarkedTexts.append(resolveStyleForMarkedText(markedTexts[0], baseStyle, renderer, lineStyle, paintInfo));
```
is this change needed?

jesxilin replied:
I changed it bc when testing out diff implementations, at one point the markedTexts could've been empty which caused the browser to crash. Testing now with the original uncheckedAppend and it works so the change isn't needed anymore.

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,22 +143,52 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+                markedTexts.append(elem);
+        }
+        // Sort so highlights follow correct priority/insertion order
+        if (!markedTexts[0].highlightName.isNull()) {
```
```suggestion
        if (!markedTexts.first().highlightName.isNull()) {
```

pxlcoder commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -103,5 +103,11 @@ void Highlight::repaint()
 
+void Highlight::setPriority(int priority)
+{
+    if (priority >= 0)
```
Is it really valid to exclude negative numbers?

The [spec](https://webidl.spec.whatwg.org/#idl-long) defines the attribute as:

```
attribute long priority;
```

Note the lack of `unsigned`.

jesxilin replied:
You're right, just found "Left the priority attribute as (implicitly) 'signed'." Will change to support negative integers.

rr-codes commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,6 +40,8 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
 void HighlightRegister::setFromMapLike(AtomString&& key, Ref<Highlight>&& value)
 {
+    AtomString k = key;
+    m_set.add(k);
```
I think this can use `WTFMove`?

nt1m commented on LayoutTests/highlight/highlight-priority.html:
This should be a web platform test.

rr-codes commented on Source/WebCore/Modules/highlight/Highlight.h:
```diff
@@ -73,7 +73,7 @@ class Highlight : public RefCounted<Highlight> {
 
     int priority() const { return m_priority; }
-    void setPriority(int priority) { m_priority = priority; }
+    void setPriority(int priority) { m_priority = priority; };
```
Extra semicolon

rr-codes replied:
Also should this be long instead of int

aprotyas commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,25 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
+            m_orderedHighlightNames.remove(i);
+            break;
+        }
+    }
```
nit: You can replace this loop with `m_orderedHighlightNames.removeFirst(key)`

aprotyas commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,38 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                        } else {
+                            // find the first correct place to insert highlight
+                            bool inserted = false;
+                            for (int8_t index = markedTexts.size()-1; index >= 0; index--) {
```
Just use `int`?

pxlcoder replied:
~I think `auto` would be better.~

aprotyas replied:
`auto` would make it unsigned, right? Would cause underflow then

pxlcoder replied:
Good point!

aprotyas commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -103,5 +103,6 @@ void Highlight::repaint()
         repaintRange(data->range());
 }
 
+
```
Nit: don’t need this new line

aprotyas commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,20 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
+    m_orderedHighlightNames.clear();
 }
 
 bool HighlightRegister::remove(const AtomString& key)
```
Do we need to assert that the `key` was actually removed? (I don't know, just a question)

jesxilin replied:
I'm not sure

megangardner replied:
I don't think so, if you try to remove a key from a map and it's not there, it should just do nothing, but we can look at the spec for map/maplike to be sure.

nt1m commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +143,42 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+    if (!markedTexts.isEmpty()) {
+        // Get rid of duplicates in vector
+        ListHashSet<MarkedText> noDuplicates;
+        for (auto& elem : markedTexts)
```
```suggestion
        for (auto& element : markedTexts)
```

Here and throughout the file, please don't use abbreviations.

jesxilin replied:
Got it!

nt1m commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +143,42 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        if (!markedTexts.first().highlightName.isNull()) {
+            // Maintain copy of old unordered markedTexts
+            Vector<MarkedText> copyOfMarkedTexts;
+            copyOfMarkedTexts = markedTexts;
```
This can be a single assignment.
Also, instead of naming this `copyOfMarkedTexts`, consider `unorderedMarkedTexts`?

aprotyas commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -368,6 +368,20 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
+        bool isHighlights = false;
+
+        for (auto& element : coalescedStyledMarkedTexts) {
+            if (!isHighlights && element.type == 6) {
```
The `!isHighlights` check isn't necessary, right? You're breaking out as soon as you set it anyway.

```suggestion
            if (element.type == MarkedText::Type::Highlight) {
```

aprotyas commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -368,6 +368,20 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
+            auto originalText = coalescedStyledMarkedTexts[coalescedStyledMarkedTexts.size()-1];
+            coalescedStyledMarkedTexts.remove(coalescedStyledMarkedTexts.size()-1);
+            coalescedStyledMarkedTexts.insert(0, originalText);
+        }
```
You could use an init statement here

```suggestion
        if (auto originalText = coalescedStyledMarkedTexts.last(); isHighlights && originalText.highlightName.isNull()) {
            coalescedStyledMarkedTexts.removeLast();
            coalescedStyledMarkedTexts.insert(0, originalText);
        }
```

aprotyas commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +167,58 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        ListHashSet<MarkedText> noDuplicates;
+        for (auto& element : markedTexts) {
+            noDuplicates.add(element);
+            if (!isHighlights && element.type == 6)
```
```suggestion
            if (!isHighlights && element.type == MarkedText::Type::Highlight)
```

aprotyas commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +167,58 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        if (!noDuplicates.isEmpty()) {
+            markedTexts.clear();
+            for (auto& element : noDuplicates)
+                markedTexts.append(element);
```
optional nit

```suggestion
            markedTexts.appendRange(noDuplicates.begin(), noDuplicates.end());
```

aprotyas commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +167,58 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        frontmostMarkedTexts.reserveInitialCapacity(markedTexts.size());
+        frontmostMarkedTexts.uncheckedAppend(resolveStyleForMarkedText(markedTexts[0], baseStyle, renderer, lineStyle, paintInfo));
+        for (auto it = markedTexts.begin() + 1, end = markedTexts.end(); it != end; ++it)
+            frontmostMarkedTexts.uncheckedAppend(resolveStyleForMarkedText(*it, baseStyle, renderer, lineStyle, paintInfo));
```
Hmm.. it isn't obvious to me why the first marked text is treated differently?

aprotyas replied:
This looks like we could get away with

```suggestion
forEach(markedTexts, [&](auto& markedText) {
    frontmostMarkedTexts.uncheckedAppend(resolveStyleForMarkedText(markedText, baseStyle, renderer, lineStyle, paintInfo));
});
```

jesxilin replied:
Yes that's correct. It's not needed anymore. Before we would compare stuff while looping but now we are just appending everything so we do not need to have an element in the vector before looping.

megangardner commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -104,4 +104,3 @@ void Highlight::repaint()
 }
 
 } // namespace WebCore
-
```
Remove this stray white space removal.

megangardner commented on Source/WebCore/Modules/highlight/Highlight.h:
```diff
@@ -84,7 +84,7 @@ class Highlight : public RefCounted<Highlight> {
     Vector<Ref<HighlightRangeData>> m_rangesData;
     Type m_type { Type::Highlight };
-    int m_priority { 0 };
+    long m_priority { 0 };
```
I'm not saying it's wrong to have done, but why did this get changed to long?

jesxilin replied:
Chris had it has `int` before, Richard pointed out that it is `long` in the spec so it should be changed. Thus, I changed it. 
Edit: it should be int, the change will be reflected in the latest PR

whsieh replied:
A `long` in C++ is a 64-bit integer on our architectures, but a `long` in IDL (confusingly) refers to a [32 bit integer](https://www.w3.org/TR/WebIDL-1/#idl-long), so I _suspect_ that `int` was correct.

jesxilin replied:
Abrar said that as well

megangardner commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -72,6 +72,7 @@ Vector<MarkedText> MarkedText::subdivide(const Vector<MarkedText>& markedTexts,
         || (a.value == b.value && a.kind == b.kind && a.kind == Offset::End && a.markedText->type > b.markedText->type);
     });
 
+
```
Remove double whitespace.

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,59 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+    for (auto it = styledTexts.begin() + 1, end = styledTexts.end(); it != end; ++it) {
+        StyledMarkedText& previousStyledMarkedText = frontmostMarkedTexts.last();
+        if (previousStyledMarkedText.startOffset == it->startOffset && previousStyledMarkedText.endOffset == it->endOffset) {
+            // Marked texts completely cover each other
```
Move this comment to above the if statement.

megangardner commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -369,6 +369,23 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
+            }
+        }
+        // Last element is original text with default color, should paint first underneath all marked texts
+        if (auto originalText = coalescedStyledMarkedTexts.last(); isHighlights && originalText.highlightName.isNull()) {
```
Should we only be doing this for the highlight case? Is it always the last element?

jesxilin replied:
I'm not sure what other case would apply? From what I tested, it is the last element and spans the range of the entire line of text. I haven't encountered a case where it is not the last element but I'll test it again.

jesxilin replied:
In `StyledMarkedText::subdivideAndResolve` I make it so any markedText that does not have a highlightName will be at the end of the vector. Since every single line of text on a webpage is a markedText that does not have a highlightName, they will always be at the end of the vector.

megangardner commented on Source/WebCore/Modules/highlight/HighlightRegister.h:
```diff
@@ -55,10 +55,12 @@ class HighlightRegister : public RefCounted<HighlightRegister> {
     
     WEBCORE_EXPORT void addAnnotationHighlightWithRange(Ref<StaticRange>&&);
     const HashMap<AtomString, Ref<Highlight>>& map() const { return m_map; }
+    const Vector<AtomString>& vector() const { return m_orderedHighlightNames; }
```
This should have a more descriptive name than just 'vector', especially since it's an added bit of implementation detail different from he spec. I suggest orderedHighlightNames()

megangardner commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +114,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                            }
+                            // insert at front of vector if lower priority than everything in markedTexts
+                            if (!inserted)
+                                markedTexts.insert(0, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
```
Do you know how often we're inserting at the front?

jesxilin replied:
In the worst case, if everything in the vector is in descending priority, we would have to keep inserting at the front for every element (which is not great).

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,59 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+{
+    // Maintain copy of old unordered markedTexts
+    Vector<MarkedText> unorderedMarkedTexts = markedTexts;
+    markedTexts.clear();
```
We don't want to clear and re-add everything. We can create the vector of highlights and insert it at the beginning, or we could consider changing this to a deque, depending on how often we need to access the beginning and the end of the structure.

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,59 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+    markedTexts.clear();
+    // Append all highlights first, in order
+    for (auto& highlightName : markedTextsNames) {
+        for (auto& element : unorderedMarkedTexts) {
```
element seems an odd choice for the variable name here, since it's a markedText

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,59 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+    }
+    // Append selection last, to maintain order
+    for (auto& element : unorderedMarkedTexts) {
+        if (element.highlightName.isNull())
```
Same as above, not sure why this is an element, and I'm not sure why you're checking if it's got a null highlight name?

jesxilin replied:
When the markedText does not have a highlightName, it is usually an indication that it is a selection markedText or essentially not a custom highlight.

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +186,42 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+
+    if (!markedTexts.isEmpty()) {
+        // Get rid of duplicates in vector, check if contains highlights
+        ListHashSet<MarkedText> noDuplicates;
```
are duplicated common? How do they end up getting added in the first place?

jesxilin replied:
Duplicates were very common and they are a result of the `MarkedText::subdivide` call

megangardner commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,11 +186,42 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        // Add back to markedTexts
+        if (!noDuplicates.isEmpty()) {
+            markedTexts.clear();
+            markedTexts.appendRange(noDuplicates.begin(), noDuplicates.end());
```
I guess I'm not sure what we're doing here, pushing the duplicated to the back of the vector?

jesxilin replied:
Before, there would be a lot of duplicate highlights with the same ranges, which would cause problems in this case (lower priority, higher priority, duplicate lower priority). Thus, to combat this I weed out the duplicate highlight and ranges to ensure we aren't going to repaint over the expected behavior. However, I'll check again if my latest changes affect this.

jesxilin replied:
I just checked and there are still duplicates. However, since I order them afterwards, having duplicates should not affect the painting order anymore so it should still work regardless of getting rid of duplicates or not.

whsieh commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +186,49 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+                isHighlights = true;
+        }
+        // Add back to markedTexts
+        if (!noDuplicates.isEmpty()) {
```
Nit - I think this `!noDuplicates.isEmpty()` check is unnecessary, since we only get here if `markedTexts` is not empty, so we'll add 1 entry to `noDuplicates` above.

whsieh commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +186,49 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+            markedTexts.clear();
+            markedTexts.appendRange(noDuplicates.begin(), noDuplicates.end());
+        }
+        // Sort so highlights follow correct priority/insertion order
```
Nit - [comments should end with a period](https://webkit.org/code-style-guidelines/#comments-sentences).

aprotyas commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -48,6 +48,13 @@ static void repaintRange(const AbstractRange& range)
     }
 }
 
+static bool repaintIfPriorityChanged(int& oldPriority, int& newPriority)
```
`int` is a POD type, just pass by value

```suggestion
static bool repaintIfPriorityChanged(int oldPriority, int newPriority)
```

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +123,82 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+static Vector<StyledMarkedText> coalesceAdjacentWithSameRanges(const Vector<StyledMarkedText>& styledTexts)
+{
+    Vector<StyledMarkedText> frontmostMarkedTexts;
+    frontmostMarkedTexts.append(styledTexts[0]);
```
is it possible for `styledTexts` to be empty?

jesxilin replied:
Don't think so, I set isHighlights to true if markedTexts is not empty and contains highlights. Thus, if it gets to this function call, it is because isHighlights is true. But to be safe I could add another check within the function, what do you think?

rr-codes replied:
I think either an `ASSERT(!styledTexts.isEmpty)` or a `if (styledTexts.isEmpty) { ASSERT_NOT_REACHED(); return { } }` would be good, just in case!

aprotyas commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -48,6 +48,13 @@ static void repaintRange(const AbstractRange& range)
+{
+    if (oldPriority != newPriority)
+        return true;
+    return false;
```
```suggestion
    return oldPriority != newPriority;
```

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +123,82 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+    }
+
+    index = 0;
+    while (index < markedTexts.size() - 1) {
```
can this just use `std::sort` instead? (not sure if it can or not)

jesxilin replied:
I only want to sort given they are highlights, adjacent and cover same ranges. There are non-highlights interspersed in the vector. Is sort able to ignore those? If so, it would be nice to change it to a one-liner.

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +123,82 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+
+    HashMap<AtomString, int> markedTextsNamesPriority;
+    unsigned index = 0;
+    for (auto& highlightName : markedTextsNames) {
```
this can probably just be a normal for loop

jesxilin replied:
Even though it is a ListHashSet? I thought they didn't have indices.

rr-codes replied:
ah, my bad!

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,43 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+            frontmostMarkedTexts.uncheckedAppend(resolveStyleForMarkedText(markedText, baseStyle, renderer, lineStyle, paintInfo));
+        });
+
+        frontmostMarkedTexts = coalesceAdjacentWithSameRanges(frontmostMarkedTexts);
```
you can just
```
return coalesceAdjacentWithSameRanges(frontmostMarkedTexts);
```
instead

rr-codes commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -366,9 +366,24 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
         }
     } else {
+        // Check if markedTexts vector contain highlights
+        bool isHighlights = false;
```
this can just use `containsIf` in `Vector.h`

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -203,4 +302,22 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
+    }
+    // Check for non-highlight StyledMarkedTexts that overlap, if so then remove them.
+    unsigned index = 0;
+    while (index < markedTexts.size()) {
```
this can be a normal for loop

jesxilin replied:
I don't increment the index if I remove something though, can it still be a normal for loop?

rr-codes replied:
ah, my bad again

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -203,4 +302,22 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
 
+void StyledMarkedText::removeOverlappingForegroundRanges(Vector<StyledMarkedText>& markedTexts)
+{
+    Vector<std::pair<unsigned, unsigned>> ranges;
```
can this be a set instead of a vector?

jesxilin replied:
I think so! Do I need to do anything special for `ListHashSet` to get it to work with a pair?

rr-codes replied:
It could probably just be a HashSet, since this one doesn't need to be ordered, right? And I don't think you'd need to do anything special, assuming our HashSet is smart enough to handle tuples

jesxilin replied:
Yes, it doesn't need to be ordered. I'll look into using HashSet now!

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,43 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+            }
+        }
+        // Sort custom highlights to follow correct priority/insertion order.
+        if (isHighlights)
```
is this supposed to be inside the `if (!markedTexts.isEmpty())` branch?

rr-codes commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,63 @@ struct MarkedText {
+        return startOffset == other.startOffset && endOffset == other.endOffset && type == other.type && marker == other.marker && priority == other.priority
+        && ((highlightName == other.highlightName) || (highlightName.isNull() && other.highlightName.isNull()));
     }
+    static bool equal(const MarkedText& a, const MarkedText& b)
```
don't think this method is needed anymore

nt1m commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,59 @@ struct MarkedText {
     {
-        return startOffset == other.startOffset && endOffset == other.endOffset && type == other.type && marker == other.marker && highlightName == other.highlightName;
+        return startOffset == other.startOffset && endOffset == other.endOffset && type == other.type && marker == other.marker && priority == other.priority
+        && ((highlightName == other.highlightName) || (highlightName.isNull() && other.highlightName.isNull()));
```
@cdumez Is the equality check enough here? or do we also need the or with both null strings? Also makes me wonder if we can use the default C++ equality operator here.

cdumez replied:
Tim is right here, you don't need these null checks. `string1 == string2` will return true if both strings are null.

nt1m commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,38 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
     if (UNLIKELY(markedTexts.isEmpty()))
         return { };
 
+    bool isHighlights = false;
```
I would name this `containsHighlight`
```suggestion
    bool containsHighlight = false;
```

nt1m commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+
+                    if (highlightStart < highlightEnd) {
+                        int currentPriority = highlightRegister->map().get(highlightName)->priority();
+                        // if we can just append it to the end, do that instead
```
nit: here and below, WebKit comment style is:
```suggestion
                        // If we can just append it to the end, do that instead.
```

nt1m commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                        } else {
+                            // find the first correct place to insert highlight
+                            bool inserted = false;
+                            for (int index = markedTexts.size()-1; index >= 0; index--) {
```
```suggestion
                            for (int index = markedTexts.size() - 1; index >= 0; index--) {
```

nt1m commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,59 @@ struct MarkedText {
+        return key.startOffset ^ key.endOffset ^ static_cast<uint8_t>(key.type);
     }
 
+
```
nit: extraneous line

nt1m commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,83 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+            && markedTextsNamesPriority.get(markedTexts[index].highlightName) > markedTextsNamesPriority.get(markedTexts[index+1].highlightName)
+            && markedTexts[index].startOffset == markedTexts[index+1].startOffset
+            && markedTexts[index].endOffset == markedTexts[index+1].endOffset) {
+            std::swap(markedTexts[index], markedTexts[index+1]);
```
```suggestion
            && !markedTexts[index + 1].highlightName.isNull()
            && markedTextsNamesPriority.get(markedTexts[index].highlightName) > markedTextsNamesPriority.get(markedTexts[index + 1].highlightName)
            && markedTexts[index].startOffset == markedTexts[index + 1].startOffset
            && markedTexts[index].endOffset == markedTexts[index + 1].endOffset) {
            std::swap(markedTexts[index], markedTexts[index + 1]);
```

rr-codes commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -48,6 +48,11 @@ static void repaintRange(const AbstractRange& range)
     }
 }
 
+static bool repaintIfPriorityChanged(int oldPriority, int newPriority)
```
Does this actually need to be a function? As far as I can tell its only used in one place

jesxilin replied:
I can put the logic in the place that is used instead of in a function. Is that preferred?

rr-codes replied:
It is, since the implementation is just an equality check

rr-codes commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,38 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+
+    if (!markedTexts.isEmpty()) {
+        // Check if vector contains custom highlights.
+        containsHighlights = markedTexts.containsIf([](const auto& item) {
```
the variable can just be both declared and defined here, right?

jesxilin replied:
yes, it was outside bc I was accessing it outside before, but I changed it now.

cdumez commented on LayoutTests/TestExpectations:
```diff
@@ -4906,7 +4906,7 @@ webkit.org/b/220325 http/wpt/css/css-highlight-api/highlight-text-cascade.html [
 imported/w3c/web-platform-tests/css/css-highlight-api/painting/css-target-text-decoration-001.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-002.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-004-2.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-004-2.html [ Pass ]
```
Why are marking these tests as Pass instead of just removing them from TestExpectations?

We only marked tests as Pass in TestExpectations when they are skipped globally and passing only on a single platform.

jesxilin replied:
I did not know that we didn't have to mark them as pass and could just remove. Will do that now!

cdumez commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -103,5 +103,13 @@ void Highlight::repaint()
-} // namespace WebCore
+void Highlight::setPriority(int priority)
+{
+    bool shouldRepaint = (m_priority != priority) ? true : false;
```
This function should be:
```
if (m_priority == priority)
    return;
m_priority = priority;
repaint();
```

cdumez commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,20 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
 
 void HighlightRegister::setFromMapLike(AtomString&& key, Ref<Highlight>&& value)
 {
+    AtomString k = key;
```
Not needed.

cdumez commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,20 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
 void HighlightRegister::setFromMapLike(AtomString&& key, Ref<Highlight>&& value)
 {
+    AtomString k = key;
+    m_orderedHighlightNames.append(WTFMove(k));
```
Just use `m_orderedHighlightNames.append(key);` here.

cdumez replied:
Here we are appending, there is no particular sorting / ordering besides the insertion order. I think we can just rename the vector to `m_highlightNames`.

Also, do we need to worry about duplicates here? Should we only append to the vector is the key isn't already in the map?

jesxilin replied:
Yes, we do need to worry about duplicates. I forgot to add that check bc it used to be a `ListHashSet` but then I changed it to a vector. Will add that check now!

cdumez commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                        int currentPriority = highlightRegister->map().get(highlightName)->priority();
+                        // If we can just append it to the end, do that instead.
+                        if (markedTexts.isEmpty()
+                            || markedTexts.last().priority <= currentPriority) {
```
nit: Could be on the previous line.

cdumez commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,39 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                            markedTexts.append({ highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
+                        } else {
+                            // Find the first correct place to insert highlight.
+                            bool inserted = false;
```
Boolean variables need a prefix as per WebKit coding style. Should be something like `wasInserted`.

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,58 @@ struct MarkedText {
     AtomString highlightName { };
-
+    int priority { 0 };
     bool isEmpty() const { return endOffset <= startOffset; }
```
First functions, then data members. Let's not mix them.

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,58 @@ struct MarkedText {
+    }
+    static unsigned hash(const MarkedText& key)
+    {
+        return key.startOffset ^ key.endOffset ^ static_cast<uint8_t>(key.type);
```
We don't need this function, see comments below.

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,58 @@ struct MarkedText {
 
 }
+namespace WTF {
+
```
I would `#include <wtf/Hasher.h>` and then add this free function for hashing:
```
inline void add(Hasher& hasher, const WebCore::Markedtext& text)
{
    add(hasher, text.startOffset, text.endOffset, text.type);
}
```

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,58 @@ struct MarkedText {
+template<> struct DefaultHash<WebCore::MarkedText> {
+    static unsigned hash(const WebCore::MarkedText& key)
+    {
+        return WebCore::MarkedText::hash(key);
```
`return computeHash(key);` now that Hasher knows how to hash this type.

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +68,58 @@ struct MarkedText {
     static Vector<MarkedText> collectForDraggedContent(const RenderText&, const TextBoxSelectableRange&);
+
+private:
+    bool m_isHashTableDeletedValue { false };
```
To avoid having an extra data member, we could just store `std::numeric_limits<unsigned>::max()` in startOffset as special value when it is the HashTableDeleted value. This is commonly done in WebKit.

cdumez commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,20 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
 
 void HighlightRegister::setFromMapLike(AtomString&& key, Ref<Highlight>&& value)
 {
+    if (!m_highlightNames.contains(key))
```
This does a linear search.

cdumez commented on Source/WebCore/Modules/highlight/HighlightRegister.cpp:
```diff
@@ -40,16 +40,20 @@ void HighlightRegister::initializeMapLike(DOMMapAdapter& map)
 {
+    if (!m_highlightNames.contains(key))
+        m_highlightNames.append(key);
     m_map.set(WTFMove(key), WTFMove(value));
```
The following would be more efficient:
```
auto addResult = m_map.set(WTFMove(key), WTFMove(value));
if (addResult.isNewEntry) {
    ASSERT(!m_highlightNames.contains(key));
    m_highlightNames.append(key);
}
```

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +69,53 @@ struct MarkedText {
-    AtomString highlightName { };
+    MarkedText(unsigned startOffset, unsigned endOffset, Type type, const RenderedDocumentMarker* marker = nullptr, AtomString highlightName = { }, int priority = 0) : startOffset(startOffset), endOffset(endOffset), type(type), marker(marker), highlightName(highlightName), priority(priority) { };
+    MarkedText(WTF::HashTableDeletedValueType) : startOffset(std::numeric_limits<unsigned>::max()) { };
+    MarkedText() { };
```
`MarkedText() = default;` would make it clear to the compiler this is still the default constructor implementation, not a custom one.

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,36 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        if (containsHighlights) {
+            orderHighlights(markedTextsNames, markedTexts);
+
+            frontmostMarkedTexts.reserveInitialCapacity(markedTexts.size());
```
I think this can be done more concisely with WTF::map():
```
frontmostMarkedTexts = WTF::map(markedTexts, [&](auto& markedTexts) {
    return resolveStyleForMarkedText(markedText, baseStyle, renderer, lineStyle, paintInfo);
});
```

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +69,53 @@ struct MarkedText {
+    unsigned endOffset { 0 };
+    Type type { Type::Unmarked };
+    const RenderedDocumentMarker* marker { nullptr };
+    AtomString highlightName { };
```
Nit: I don't think you need `{ }`. Please check if the compiler is happy without it.

jesxilin replied:
the complier is happy!

cdumez commented on Source/WebCore/rendering/MarkedText.h:
```diff
@@ -68,24 +69,53 @@ struct MarkedText {
-    Type type;
-    const RenderedDocumentMarker* marker { nullptr };
-    AtomString highlightName { };
+    MarkedText(unsigned startOffset, unsigned endOffset, Type type, const RenderedDocumentMarker* marker = nullptr, AtomString highlightName = { }, int priority = 0) : startOffset(startOffset), endOffset(endOffset), type(type), marker(marker), highlightName(highlightName), priority(priority) { };
```
We should not be passed AtomString by value, this causes so RefCounting churn internally when copying the AtomString. Should use `const AtomString&` or `AtomString&&` if the call sites can `WTFMove()` it in.

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,83 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+        return;
+
+    HashMap<AtomString, int> markedTextsNamesPriority;
+    unsigned index = 0;
```
nit: Seems we could use `int` since that's how you store it in the HashMap anyway.

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,83 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
     return style;
 }
 
+static Vector<StyledMarkedText> coalesceAdjacentWithSameRanges(const Vector<StyledMarkedText>& styledTexts)
```
We can probably take the Vector as a r-value reference for perf:
`Vector<StyledMarkedText>&&`

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,83 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+            }
+            continue;
+        }
+        frontmostMarkedTexts.append(*it);
```
Then you could `WTFMove(*it)` here.

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -120,11 +122,83 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
+            markedTextsNames.add(markedText.highlightName);
+    }
+
     Vector<StyledMarkedText> frontmostMarkedTexts;
```
Can we drop this...

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,35 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+        if (containsHighlights) {
+            orderHighlights(markedTextsNames, markedTexts);
+
+            frontmostMarkedTexts = WTF::map(markedTexts, [&](auto& markedText) {
```
... and use `auto` here? It seems `frontmostMarkedTexts` is not used elsewhere.

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -136,18 +210,35 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
+            frontmostMarkedTexts = WTF::map(markedTexts, [&](auto& markedText) {
+                return resolveStyleForMarkedText(markedText, baseStyle, renderer, lineStyle, paintInfo);
+            });
+            return coalesceAdjacentWithSameRanges(frontmostMarkedTexts);
```
`WTFMove(frontmostMarkedTexts)`

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -203,4 +294,22 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
+            markedTexts.remove(index);
+            continue;
+        }
+        index++;
```
Although it doesn't matter for built-in types, it is good practice to use pre-increment / pre-decrement whenever possible (instead of the postfix version).

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -203,4 +294,22 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
+    unsigned index = 0;
+    while (index < markedTexts.size()) {
+        if (markedTexts[index].highlightName.isNull() && ranges.contains(std::make_pair(markedTexts[index].startOffset, markedTexts[index].endOffset))) {
+            markedTexts.remove(index);
```
Seems like we may be able to simplify this code a bit if you used `Vector::removeAllMatching()`

cdumez reviewed (commented):
Looks good overall from a C++/Style standpoint. Still a few minor comments / suggestions.
Would be good for a CSS person to take a look too.

cdumez commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -366,9 +366,25 @@ void TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations()
+        if (coalescedStyledMarkedTexts.isEmpty())
+            return;
+
+        // DEBUG JESS
```
I guess you forgot to remove this :)

jesxilin replied:
Added that debugging code bc a test only fails on linux rn :(

cdumez commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -203,4 +299,17 @@ Vector<StyledMarkedText> StyledMarkedText::coalesceAdjacentWithEqualDecorations(
     });
 }
 
+void StyledMarkedText::removeOverlappingForegroundRanges(Vector<StyledMarkedText>& markedTexts)
```
I may be misunderstanding but the code in this function doesn't seem to make sense. It seems:
1. You create a HashSet containing all the [startOffset, endOffset] of the markedTexts that don't have a highlightName
2. Then you remove from markedTexts all the texts that have don't have a highlightName and whose [startOffset, endOffset] is in the HashSet.

It seems it would remove all markedTexts that don't have a highlight name no?

cdumez replied:
If I understand correctly the purpose, maybe you could write it like so:
```
HashSet<std::pair<unsigned, unsigned>> ranges;
markedTexts.removeAllMatching([&] (auto& markedText) {
    return markedText.highlightName.isNull() && !ranges.add(std::pair { markedText.startOffset, markedText.endOffset }).isNewEntry;
});
```

cdumez replied:
If 2 ranges overlap, I am not sure if you want to remove all of them or only the second one. My code above assumes you want to remove all that overlaps except for the first one.

If you want to remove all the ranges that overlap, you could do something like:
```
HashCountedSet<std::pair<unsigned, unsigned>> ranges;
for (auto& markedText : markedTexts) {
    if (!markedText.highlightName.isNull())
        ranges.add(std::pair { markedText.startOffset, markedText.endOffset });
}
markedTexts.removeAllMatching([&](auto& markedText) {
    return markedText.highlightName.isNull() && ranges.count(std::pair { markedText.startOffset, markedText.endOffset }) > 1;
});
```

Either way, removing all markedText which doesn't have a highlight name doesn't seem what you intended to do based on your comment inside the function where you talk about overlap.

Also note that your comment mentions "overlap", I think 2 ranges can have overlap without necessarily be identical.
I didn't check the spec but it'd be good to check what the exact behavior is supposed to be. Maybe we want to merge overlapping ranges into a single range by doing a union for e.g.

jesxilin replied:
I have a `HashSet` containing all ranges of `MarkedTexts` that _do_ have a highlight name (aka highlight ranges). Then, I remove all the MarkedText that are identical with those specific ranges. This is because the vector contains MarkedTexts that are the original text and MarkedTexts of highlights that contain the word the highlight covers. Thus, it would paint the text of the highlight and then another `MarkedText` with identical range would paint over that same word. This caused unwanted behavior if the custom highlight specified a different text color. At the time, I did not have a sophisticated coalescing styles function and I sorted the vector so all non-highlights would be at the end of the vector. Due to the sorting, I would never coalesce styles of all the `MarkedTexts` that had identical ranges since they were not adjacent to each other in vector. So, my initial idea was to remove overlapping `MarkedTexts` that had identical ranges as highlights. However, just after testing, this function doesn't do anything anymore because I improved my "coalescing adjacent identical ranges" function and took out the sorting. Sorry for the confusion, I'll remove this piece of code for the next push.

jesxilin replied:
If what I said doesn't make any sense, pls let me know and I can explain better with visuals because the way highlights works with the existing code rn is definitely confusing 😞

megangardner commented on Source/WebCore/rendering/MarkedText.cpp:
```diff
@@ -113,19 +113,38 @@ Vector<MarkedText> MarkedText::collectForHighlights(const RenderText& renderer,
+                                }
+                            }
+                            // Insert at front of vector if lower priority than everything in markedTexts.
+                            if (!wasInserted)
```
I really want this to be added into the above loop, but I'm not sure how to do it cleanly. But having the same insert line can lead to bugs in the future, so we should see if there's a nice way to combine these. But that can be in a follow up.

aprotyas replied:
Hmm... it seems like `markedTexts` is sorted in ascending order.

If so, couldn't we do this?

```cpp
// Finds iterator of the first markedText with priority < currentPriority
auto it = std::lower_bound(markedTexts.begin(), markedTexts.end(), key, [](const auto& m1, const auto& m2) {
    return m1.priority < m2.priority;
});

unsigned insertIndex = (it == markedTexts.end() ? 0 : std::distance(markedTexts.begin(), it) + 1);
markedTexts.insert(insertIndex, { highlightStart, highlightEnd, MarkedText::Type::Highlight, nullptr, highlightName, currentPriority });
```

jesxilin replied:
above loop as in `for (auto& highlightName : highlightRegister->highlightNames())`?

jesxilin committed:
Add priority for Highlight
https://bugs.webkit.org/show_bug.cgi?id=257611
rdar://110125853

Reviewed by Megan Gardner.

Included implementation for priority and insertion order of css highlights.
Fixed selection inconsistency to always have active selection painted the same, above all markedTexts.
Fixed painting foreground color text so non-clashing styles of lower priority are still painted.
Added implementation for repainting when priority changes.
Added implementation for painting same semi-transparent overlapping highlight to be painted as a union.
Added a check if users assigned text color to take text color of prioritized highlight.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority-painting.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-highlight-api/highlight-priority.html: Added.
* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::Highlight::setPriority):
* Source/WebCore/Modules/highlight/Highlight.h:
(WebCore::Highlight::setPriority): Deleted.
* Source/WebCore/Modules/highlight/HighlightRegister.cpp:
(WebCore::HighlightRegister::setFromMapLike):
(WebCore::HighlightRegister::clear):
(WebCore::HighlightRegister::remove):
* Source/WebCore/Modules/highlight/HighlightRegister.h:
(WebCore::HighlightRegister::highlightNames const):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::subdivide):
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/MarkedText.h:
(WebCore::MarkedText::MarkedText):
(WebCore::MarkedText::isHashTableDeletedValue const):
(WTF::add):
(WTF::HashTraits<WebCore::MarkedText>::constructDeletedValue):
(WTF::HashTraits<WebCore::MarkedText>::isDeletedValue):
(WTF::DefaultHash<WebCore::MarkedText>::hash):
(WTF::DefaultHash<WebCore::MarkedText>::equal):
(WebCore::MarkedText::operator== const): Deleted.
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
(WebCore::coalesceAdjacentWithSameRanges):
(WebCore::orderHighlights):
(WebCore::StyledMarkedText::subdivideAndResolve):
(WebCore::StyledMarkedText::coalesceAdjacentWithEqualDecorations):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintForegroundAndDecorations):
* Source/WebCore/rendering/TextPaintStyle.h:

Canonical link: https://commits.webkit.org/265812@main
